src/activation.rs

```diff
@@ -0,0 +1,155 @@
+use wayland_client::{
+    globals::{BindError, GlobalList},
+    protocol::{wl_seat, wl_surface},
+    Dispatch, Proxy, QueueHandle,
+};
+use wayland_protocols::xdg::activation::v1::client::{xdg_activation_token_v1, xdg_activation_v1};
+
+use crate::{
+    error::GlobalError,
+    globals::{GlobalData, ProvidesBoundGlobal},
+};
+
+/// Data attached to a token request
+pub trait RequestData: Send + Sync {
+    /// App_id of the application requesting the token, if applicable
+    fn app_id(&self) -> Option<String>;
+    /// Seat and serial of the window requesting the token, if applicable.
+    ///
+    /// *Warning*: Many compositors will issue invalid tokens for requests without
+    /// recent serials. There is no way to detect this from the client-side.
+    fn serial(&self) -> Option<(&wl_seat::WlSeat, u32)>;
```

Comment:
You're returning `seat` and `serial` from a function called `serial`, sounds really odd.

I'd suggest to separate `seat` and `serial`. Also `&Wlseat` should be better.

Reply:
That's the naming the protocol uses and it expects either both or none of the above. So I can't really make two methods out of it. We *could* just change the name.

Reply:
this better?

Reply:
Well, if you tell me what the implementation is supposed to do, if it gets either but not the other... just don't set it? That feels wrong, if we can enforce it on a type level.